### PR TITLE
Simplify position arithmetics in AbstractSingleRowBlock

### PR DESCRIPTION
**Pull request by findepi**

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/block/TestRowBlock.java [hunk 1]
```diff
@@ -154,6 +154,7 @@ else if (fieldTypes.get(j) == VARCHAR) {
                     }
                 }
             }
+            testRows[i] = testRow;
```
Comment (wenleix) [inline presto-main/src/test/java/com/facebook/presto/block/TestRowBlock.java]: My bad. Thank you for catching this.

**Comment by findepi** [comment]
@wenleix i rebased that to cater for #9388

File: presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractSingleRowBlock.java [hunk 2]
```diff
@@ -21,137 +21,140 @@
 public abstract class AbstractSingleRowBlock
         implements Block
 {
-    // in AbstractSingleRowBlock, offset is position-based (consider as cell-based), not entry-based.
-    protected final int startOffset;
+    protected final int rowOffset;
```
Comment (wenleix) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractSingleRowBlock.java]: Maybe say `startRowOffset` ? @haozhun any thoughts?

**Comment by wenleix** [comment]
Note this will make the offset in SingleRowBlock different with the offset in SingleArray/MapBlock (entry-based offset vs. position-based offset). However, this seems reasonable as the difference of RowBlock (each entry has fixed number of positions) vs. Array/MapBlock (each entry has variant number of positions) and how it can simplify the logic in SingleRowBlock.

Does this sound reasonable to you ? @haozhun @dain

File: presto-spi/src/main/java/com/facebook/presto/spi/block/SingleRowBlockWriter.java [hunk 3]
```diff
@@ -177,7 +179,7 @@ private void entryAdded()
     }
 
     @Override
-    public int getPositionCount()
+    public int getFieldCount()
```
[the remark this answers is not in the document]
Comment (findepi) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/SingleRowBlockWriter.java]: If this is confusing, then let's go with `abstract getPositionCount()`.

**Comment by findepi** [comment]
> Note this will make the offset in SingleRowBlock different with the offset in SingleArray/MapBlock (entry-based offset vs. position-based offset).

That's probably the mian point of this PR. Having Row, Map and Array blocks similar made Row block more complex than necessary* and at the same time no code was shared between the tree implementations. 

\* the complexity manifested in all the `getXXX(int position)` methods, where we converted `position` to `getAbsolutePosition(position)` then computed modulo `% numFields` and it was quite not obvious that the end result is nothing else than initial `position`.

**Comment by findepi** [comment]
rebased on current master

File: presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractSingleRowBlock.java (hunk 2 above)
Comment (wenleix) [reply presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractSingleRowBlock.java]: Let's call it `rowIndex` .

File: presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractSingleRowBlock.java [hunk 4]
```diff
@@ -21,137 +21,135 @@
...
 
     protected abstract Block getFieldBlock(int fieldIndex);
 
-    protected AbstractSingleRowBlock(int startOffset, int numFields)
-    {
-        this.startOffset = startOffset;
-        this.numFields = numFields;
-    }
+    @Override
+    public abstract int getPositionCount();
```
Comment (wenleix) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/AbstractSingleRowBlock.java]: This abstract method definition is redundant  (it's defined by `Block` interface)

File: presto-spi/src/main/java/com/facebook/presto/spi/block/SingleRowBlock.java [hunk 5]
```diff
@@ -77,21 +77,21 @@ public void retainedBytesForEachPart(BiConsumer<Object, Long> consumer)
...
-        for (int i = 0; i < numFields; i++) {
+        BlockEncoding[] fieldBlockEncodings = new BlockEncoding[fieldBlocks.length];
+        for (int i = 0; i < fieldBlocks.length; i++) {
             fieldBlockEncodings[i] = fieldBlocks[i].getEncoding();
         }
         return new SingleRowBlockEncoding(fieldBlockEncodings);
     }
 
-    public int getOffset()
+    public int getRowOffset()
```
Comment (wenleix) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/SingleRowBlock.java]: `getRowIndex`. Do the same for `SingleRowBlockEncoding::writeBlock`

File: presto-spi/src/main/java/com/facebook/presto/spi/block/SingleRowBlockWriter.java [hunk 6]
```diff
@@ -27,16 +27,18 @@
     private static final int INSTANCE_SIZE = ClassLayout.parseClass(SingleRowBlockWriter.class).instanceSize();
 
     private final BlockBuilder[] fieldBlockBuilders;
+    private final int numFields;
```
Comment (wenleix) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/SingleRowBlockWriter.java]: Remove this, use `fieldBlockBuilders.length` to be consistent with `SingleRowBlock`

File: presto-spi/src/main/java/com/facebook/presto/spi/block/SingleRowBlockWriter.java [hunk 7]
```diff
@@ -27,16 +27,18 @@
...
     private final long initialBlockBuilderSize;
     private int positionsWritten;
 
     private int currentFieldIndexToWrite;
     private boolean fieldBlockBuilderReturned;
 
-    SingleRowBlockWriter(int startOffset, BlockBuilder[] fieldBlockBuilders)
+    SingleRowBlockWriter(int rowOffset, BlockBuilder[] fieldBlockBuilders)
```
Comment (wenleix) [inline presto-spi/src/main/java/com/facebook/presto/spi/block/SingleRowBlockWriter.java]: `rowIndex`

**Comment by findepi** [comment]
@wenleix thanks, comments applied.

**Comment by wenleix** [comment]
Merged #9383, thank you !

**Comment by findepi** [comment]
@wenleix thanks for review!

**Pull request closed without merging**